### DISCUSSION TITLE
Cross Compiling Issue

### DISCUSSION THREAD

**Discussion by gitkarokan** [mesonbuild/meson#10412] [Q&A] [answered]
I'm trying to transition from my current build system SCons into Meson so I started with a simple project. I'm doing a cross-compilation of a simple project (5 files); it should be straight forward but I'm getting the following error when running this command `meson buildDir --cross-file .\cross_file.build`  and I hope you can provide some help or clarity why it is happening

> meson.build:1:0: ERROR: Unable to detect linker for compiler ".\tools\gcc_compiler\bin\ppc-freevle-eabi-gcc -Wl,--version -c -mcpu=e200z4 -mvle -meabi -msdata=none -mregnames -O2 -gdwarf-2 -fomit-frame-pointer -falign-functions=16 -fno-gcse -std=gnu99"

```
[constants]
compiler_path = '.\tools\gcc_compiler\bin\'
arch = 'ppc-freevle-eabi' 

[binaries]
c       = compiler_path + arch + '-gcc'
cpp     = compiler_path + arch + '-cpp'
ld      = compiler_path + arch + '-ld'
ar      = compiler_path + arch + '-ar'
as      = compiler_path + arch + '-as'
size    = compiler_path + arch + '-size'
objdump = compiler_path + arch + '-objdump'
objcopy = compiler_path + arch + '-objcopy'
strip   = compiler_path + arch + '-strip'
gdb     = compiler_path + arch + '-gdb'

[built-in options]
c_args      = [
                '-c',
                '-mcpu=e200z4',
                '-mvle',
                '-meabi',
                '-msdata=none',
                '-mregnames',
                '-O2',
                '-gdwarf-2',
                '-fomit-frame-pointer',
                '-falign-functions=16',
                '-fno-gcse',
                '-std=gnu99'
               ]

c_link_args = [
               '-mcpu=e200z4',
               '-mvle',
               '-meabi',
               '-msdata=none',
               '-mregnames',
               '-O2',
               '-gdwarf-2',
               '-fomit-frame-pointer',
               '-falign-functions=16',
               '-fno-gcse',
               '-std=gnu99',
               '-ffunction-sections',
               '-fdata-sections',
               '-fno-common',
               '-nostartfiles',
               '-Wl,-Map=build/out.map',
               '--cref,--no-warn-mismatch,--library-path=,--gc-sections,-lm,-lc,-z,max-page-size=16384,--script=application.ld -o build/out.elf'
              ]

[host_machine]
system = 'windows'
cpu_family = 'x86'
cpu = 'i686'
endian = 'little'
```

If I execute the same command in the command line `".\tools\gcc_compiler\bin\ppc-freevle-eabi-gcc -Wl,--version -c -mcpu=e200z4 -mvle -meabi -msdata=none -mregnames -O2 -gdwarf-2 -fomit-frame-pointer -falign-functions=16 -fno-gcse -std=gnu99"` and I don't get any error, any idea?

Thanks!

**Comment by eli-schwartz** [comment]
> If I execute the same command in the command line `".\tools\gcc_compiler\bin\ppc-freevle-eabi-gcc -Wl,--version -c -mcpu=e200z4 -mvle -meabi -msdata=none -mregnames -O2 -gdwarf-2 -fomit-frame-pointer -falign-functions=16 -fno-gcse -std=gnu99"` and I don't get any error, any idea?

You're not supposed to get an error, but you *are* supposed to get output, something like:
```
collect2 version 11.1.0
/usr/bin/ld -plugin /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/liblto_plugin.so -plugin-opt=/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/lto-wrapper -plugin-opt=-fresolution=/tmp/cc48PP1g.res -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s --build-id --eh-frame-hdr --hash-style=gnu -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -pie /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/Scrt1.o /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/crti.o /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/crtbeginS.o -L/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0 -L/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib -L/lib/../lib -L/usr/lib/../lib -L/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../.. --version -lgcc --push-state --as-needed -lgcc_s --pop-state -lc -lgcc --push-state --as-needed -lgcc_s --pop-state /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/crtendS.o /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/crtn.o
GNU ld (GNU Binutils) 2.36.1
Copyright (C) 2021 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.
```

However, using -c in your `c_args` seems pretty wrong and causes this problem. The purpose of this argument is to tell the compiler on a per-invocation basis "I am currently compiling a translation unit from source code to an intermediate object, and ***later*** I will link the object together with a bunch of others into one final executable/library, do NOT invoke the linker".

This breaks because then -c is *always* used for all compiler invocations, including the link stage. More specifically, it means that when running the linker detection, the linker is `-c` == not run, so we can't inspect its --version output. :(

I'm not sure what your goal is in specifying `-c`, this is supposed to be handled by meson itself for each internally generated intermediate rule.

**Comment by gitkarokan** [comment, marked as answer]
Thank you @eli-schwartz for your quick reply! 

I was using the -c to mimic exactly the build command from the IDE, thanks for the explanation. 

After removing the -c, meson still shows this error `meson.build:1:0: ERROR: Compiler .\tools\gcc_compiler\bin\ppc-freevle-eabi-gcc can not compile programs.` when running this command `meson buildDir --cross-file .\cross_file.build` 

Below is the log file 

```
Build started at 2022-05-19T19:28:55.243647
Main binary: C:\Python\python.exe
Build Options: '--cross-file .\cross_file.build'
Python system: Windows
The Meson build system
Version: 0.62.0
Source dir: xxxxxxxxxxxxxxxxxxxxxxxxxxx\03_Projects\01_Meson
Build dir: xxxxxxxxxxxxxxxxxxxxxxxxxxxxx\03_Projects\01_Meson\buildDir
Build type: cross build
Project name: Testing
Project version: undefined
Activating VS 17.1.4
Sanity testing C compiler: .\tools\gcc_compiler\bin\ppc-freevle-eabi-gcc
Is cross compiler: True.
Sanity check compiler command line: .\tools\gcc_compiler\bin\ppc-freevle-eabi-gcc sanitycheckc.c -o sanitycheckc_cross.exe -mcpu=e200z4 -mvle -meabi -msdata=none -mregnames -O2 -gdwarf-2 -fomit-frame-pointer -falign-functions=16 -fno-gcse -std=gnu99 -D_FILE_OFFSET_BITS=64 -c
Sanity check compile stdout:

-----
Sanity check compile stderr:
ppc-freevle-eabi-gcc: error: CreateProcess: No such file or directory

-----

meson.build:1:0: ERROR: Compiler .\tools\gcc_compiler\bin\ppc-freevle-eabi-gcc can not compile programs.
```

I did a test and I was able to build the binary files using this command `.\tools\gcc_compiler\bin\ppc-freevle-eabi-gcc build\meson-private\sanitycheckc.c -o build\meson-private\sanitycheckc_cross.exe -mcpu=e200z4 -mvle -meabi -msdata=none -mregnames -O2 -gdwarf-2 -fomit-frame-pointer -falign-functions=16 -fno-gcse -std=gnu99 -D_FILE_OFFSET_BITS=64 -c`

Please notice that I need to add _build\meson-private_ in front of the file.

Do I need to specify anything else on my cross-compilation file? 

Thanks a lot!

**Reply by eli-schwartz** [reply]
I'm pretty sure you need it to be an absolute path. You can either use a tiny script to create the cross file based on the checkout directory, if the tools are checked into VCS, or there is also a proposal to add the containing directory of the cross file itself as a built-in cross file constant IIRC.

**Reply by gitkarokan** [reply]
Thanks a lot @eli-schwartz for your help! This solved the issue, thanks again!
